File: python/src/hyperfly/registry.py
from __future__ import annotations

from typing import Iterable

from ._codec import Codec


class CodecRegistry:
    """Codecs keyed by fingerprint.

    Rotation is why this exists: retraining a profile produces a new fingerprint, so a
    deployment holding only one codec per route turns every rollout into a cutover in
    which in-flight clients fall back to JSON. Holding the outgoing codec alongside the
    incoming one makes that a transition instead.
    """

    def __init__(self, codecs: Iterable[Codec] = ()) -> None:
        self._by_fingerprint: dict[str, Codec] = {}
        for codec in codecs:
            self.add(codec)

    def add(self, codec: Codec) -> "CodecRegistry":
        self._by_fingerprint[codec.fingerprint] = codec
        return self

    def remove(self, fingerprint: str) -> bool:
        return self._by_fingerprint.pop(fingerprint, None) is not None

    def get(self, fingerprint: str) -> Codec | None:
        return self._by_fingerprint.get(fingerprint)

    def __contains__(self, fingerprint: object) -> bool:
        return fingerprint in self._by_fingerprint

    def __len__(self) -> int:
        return len(self._by_fingerprint)

    @property
    def fingerprints(self) -> list[str]:
        return list(self._by_fingerprint)

    def artifact(self, fingerprint: str) -> str | None:
        codec = self._by_fingerprint.get(fingerprint)
        return codec.artifact if codec else None

    def select(self, accepted: Iterable[str]) -> Codec | None:
        """The client's preference wins, which is what lets it migrate itself."""
        for fingerprint in accepted:
            codec = self._by_fingerprint.get(fingerprint)
            if codec is not None:
                return codec
        return None

File: packages/hyperfly/src/registry.ts
import type { Codec } from "./codec.js";

/**
 * Codecs keyed by fingerprint. Rotation is why this exists: retraining a profile
 * produces a new fingerprint, so a deployment that holds only one codec per route
 * turns every rollout into a cutover in which in-flight clients fall back to JSON.
 * Holding the outgoing codec alongside the incoming one makes that a transition.
 */
export class CodecRegistry {
  private readonly byFingerprint = new Map<string, Codec<never>>();

  constructor(codecs: readonly Codec<never>[] = []) {
    for (const codec of codecs) this.add(codec);
  }

  add(codec: Codec<never>): this {
    this.byFingerprint.set(codec.fingerprint, codec);
    return this;
  }

  remove(fingerprint: string): boolean {
    return this.byFingerprint.delete(fingerprint);
  }

  get(fingerprint: string): Codec<never> | undefined {
    return this.byFingerprint.get(fingerprint);
  }

  has(fingerprint: string): boolean {
    return this.byFingerprint.has(fingerprint);
  }

  get fingerprints(): string[] {
    return [...this.byFingerprint.keys()];
  }

  get size(): number {
    return this.byFingerprint.size;
  }

  /** The artifact text for a fingerprint, for `.well-known` discovery (negotiation §3). */
  artifact(fingerprint: string): string | undefined {
    return this.byFingerprint.get(fingerprint)?.artifact;
  }

  /**
   * The client's preference wins: the first fingerprint it lists that we can serve.
   * That is what lets a client migrate itself during a rotation without the server
   * tracking who holds what.
   */
  select(accepted: readonly string[]): Codec<never> | undefined {
    for (const fingerprint of accepted) {
      const codec = this.byFingerprint.get(fingerprint);
      if (codec) return codec;
    }
    return undefined;
  }
}
